js/pastFutureDateTags.js

```diff
@@ -0,0 +1,347 @@
+ *
+ *  Option 2: Change returntype = "hard", texttype = "date", regexVal = null
+ *  Pros: Proper date validation, focuses field on failed validation, no additional code
+ *  Cons: Does not maintain date field format
+ *  Note: Changing the texttype to date does not maintain the format for date_xyz fields
+ *  Sample Code: 
+ *  $input.attr("onblur", "redcap_validate(this, '2021-01-15','2025-10-01','hard','date',null)")
+ *
+ *  Option 3: Wrap the redcap_validate call and refocus element on failed validation
+ *  Pros: Can focus field on failed validation, no duplicate code, easier to maintain
+ *  Cons: Unable to override alert message that includes 'This value is admissable, but you may wish to verify.'
+ *  Sample Code:
+ *  function wrapValidation(self, innerFunction) {
+ *      var result = innerFunction(self);
+ *      console.log('returned from inner function with: ' + result);
+ *      // if the element is not valid then focus it
+ *  } 
+ *  $input.attr("onblur", "wrapValidation(this, function(self) { return redcap_validate(self,'2021-01-15','2025-10-01','hard','date_ymd',1);})")
+ *
+ *  (Implemented) Option 4: Take relevant redcap_validate code and tweak failed validation behavior and alert message
+ *  Pros: Proper date validation, overrides alert message, focuses field on failed validation, maintains date format
+ *  Cons: Duplicate code
+ * 
+ * 
+ *  REDCap alert call:
+ *  var uniqueID = 'someUniqueID';
+ *  var msg = 'some message';
+ *  initDialog(uniqueID);
+ *  $(`#${uniqueID}`).html(msg);
+ *  simpleDialog(msg, null, uniqueID);
+ * 
+ *  Relevant thread: // https://community.projectredcap.org/questions/41383/use-datepicker-in-external-module-custom-code.html
+ **/
+const reDates = /(\d{4}-\d{2}-\d{2}[^']*)/g;
+let isValidClick = false;
+
+$(document).ready(function () {
+	let module = ExternalModules['PastFutureDateTags'].ExternalModule;
+	let preventFutureDateFields = JSON.parse(module.tt('preventFutureDateFields'));
+	let preventPastDateFields = JSON.parse(module.tt('preventPastDateFields'));
+	let attachedListeners = false;
+
+	// REDCap calls $.datepicker.setDefaults() after External Module javascript is executed.
+	// Because of this the restrictions on the jQuery UI datepicker need to be applied after
+	// all scripts have been loaded.
+	$('#form').on('mousemove', function() {
+		if (!attachedListeners) {
+			attachedListeners = true;
+			applyDateRestrictions(preventFutureDateFields, preventFutureDate);
+			applyDateRestrictions(preventPastDateFields, preventPastDate);
+		}
+	});
+});
+
+function applyDateRestrictions(dateFields, fn) {
+	for (const field of dateFields) {
+		fn(field);
+	}
+}
+
+// Sets the maxDate option on the jQuery UI datepicker
+function preventFutureDate(field) {
+	let $input = $(`#${field}-tr input`);
+	let val = $input.attr("onblur");
+	let dateFormat = $input.attr('fv');
+	let [minDate, _] = val.match(reDates);
```

Comment:
`Uncaught TypeError: can't access property Symbol.iterator, val.match(...) is null`